provisioner/node_pools.go

```diff
@@ -0,0 +1,357 @@
+package provisioner
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"text/template"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando-incubator/cluster-lifecycle-manager/api"
+	awsExt "github.com/zalando-incubator/cluster-lifecycle-manager/pkg/aws"
+	"github.com/zalando-incubator/cluster-lifecycle-manager/pkg/updatestrategy"
+)
+
+const (
+	userDataFileName      = "userdata.clc.yaml"
+	stackFileName         = "stack.yaml"
+	nodePoolTagKeyLegacy  = "NodePool"
+	nodePoolTagKey        = "kubernetes.io/node-pool"
+	nodePoolRoleTagKey    = "kubernetes.io/role/node-pool"
+	nodePoolProfileTagKey = "kubernetes.io/node-pool/profile"
+)
+
+// NodePoolProvisioner is able to provision node pools for a cluster.
+type NodePoolProvisioner interface {
+	Provision(values map[string]string) error
+	Reconcile() error
+}
+
+// AWSNodePoolProvisioner is a node provisioner able to provision node pools
+// in AWS via cloudformation.
+type AWSNodePoolProvisioner struct {
+	awsAdapter      *awsAdapter
+	nodePoolManager updatestrategy.NodePoolManager
+	bucketName      string
+	cfgBaseDir      string
+	Cluster         *api.Cluster
+	logger          *log.Entry
+}
+
+// stackParams defined the parameters expected by a node pool stack template.
+type stackParams struct {
+	Cluster  *api.Cluster
+	NodePool *api.NodePool
+	UserData string
+	Values   map[string]string
+}
+
+type userDataParams struct {
+	Cluster  *api.Cluster
+	NodePool *api.NodePool
+	Values   map[string]string
+}
+
+func (n *AWSNodePoolProvisioner) generateNodePoolStackTemplate(nodePool *api.NodePool, values map[string]string) (string, error) {
```

Comment:
I would prefer `(npp *AWSNodePoolProvisioner)` or `*p`

Reply:
Done (changed to `p`)